PR: https://github.com/odoo/odoo/pull/261571

From: c03c3502040d04c26b568cd25c1368d500236544
From: Frantisek Binovsky
Date: 2026-09-15 16:55:01

Breaking data model changes scores: del:3 + add:8, change matches:
+    event_ids = fields.One2many('calendar.event', 'calendar_id', 'Events')
+    recurrence_ids = fields.One2many('calendar.recurrence', 'calendar_id', 'Recurrences')
+    calendar_user_ids = fields.One2many('calendar.user', inverse_name='calendar_id', string='Users', default=_default_calendar_user_ids)
+    share_user_ids = fields.Many2many('res.users', string='Shared with', compute='_compute_share_user_ids',
-    calendar_default_privacy = fields.Selection(
+    calendar_user_ids = fields.One2many('calendar.user', 'user_id')
+    calendar_ids = fields.One2many('calendar.calendar', compute='_compute_calendar_ids', readonly=False, user_writeable=True)
+    owned_calendar_ids = fields.One2many('calendar.calendar', compute='_compute_calendar_ids')
+    writable_calendar_ids = fields.One2many('calendar.calendar', compute='_compute_calendar_ids')
-    calendar_default_privacy = fields.Selection(
-    calendar_hide_my = fields.Boolean(string='Hide My Calendar from Calendar')

Total Changes: 493

[IMP] calendar,hr_calendar,web: introduce multiple calendars per user

Before this commit, the user model had a direct One2Many relation with events.
To organize the events into different calendars, new models were introduced,
resulting in the following structure:

* user --o2m--> calendar.user --m2o--> calendar --o2m--> event

Because Google and Outlook support sharing calendars between users, we
introduced a calendar.calendar.user 'membership' model to allow more users
to share one calendar. The model holds the user's access role to a calendar
along with any relevant metadata for each calendar/user pair, such as saved
filters and if it is the user's primary calendar

The calendar_default_privacy field was moved from user_settings to the
individual calendars.

Each user should have exactly one primary calendar, which cannot be deleted.
This is enforced by a migration script and a post_init hook which creates this
calendar for each existing user. For new users, this is handled in a 'Create'
override. A db constraint was added to ensure each user has no more than one
primary calendar.

The filter section can be used to edit/create/share calendars via buttons
shown on hover. Alternatively, a simple form/list view was added to the
configuration menu and user preferences.

The colors of the events are based on:
1. Calendars if they belong to the user's calendar
2. partner_id otherwise (just like before)

### Web changes:

Because of the introduction of multiple filters with custom labels in the
attendee calendar, the autocomplete placeholders no longer worked (for attendee
filters, the new label is `Meet with`, so the autocomplete would result in `+
Add Meet with`)
*> Added a placeholder field to the calendar filter section autocomplete

The rest of the web changes include refactoring to be able to override specific
parts of methods.

In the changed test, the 'calendar.user' model mock which refers to a res.user
was renamed so it's not confused with the newly introduced `calendar.user` model

Task-5485938

Part-of: odoo/odoo#261571
Related: odoo/enterprise#115416
Related: odoo/upgrade#10067
Signed-off-by: Warnon Aurélien (awa) <awa@odoo.com>

================================= pseudo patch: =================================

--- a/addons/calendar/models/__init__.py
+++ b/addons/calendar/models/__init__.py
@@ -2,6 +2,8 @@
 
 from . import ir_http
 from . import res_partner
+from . import calendar_calendar
+from . import calendar_user
 from . import calendar_event
 from . import calendar_alarm
 from . import calendar_alarm_manager

--- a/None
+++ b/addons/calendar/models/calendar_calendar.py
@@ -0,0 +1,144 @@
+from odoo import api, Command, fields, models, _
+from odoo.exceptions import UserError, AccessError
+
+
+class CalendarCalendar(models.Model):
+    _name = 'calendar.calendar'
+    _description = 'User Calendar'
+
+    def _default_calendar_user_ids(self):
+        return [Command.create({
+            'user_id': self.env.user.id,
+            'access_role': 'owner',
+            'is_filter_active': True,
+            'is_filter_checked': True,
+            'name': _('Calendar')
+        })]
+
+    name = fields.Char(string='Name')
+    event_ids = fields.One2many('calendar.event', 'calendar_id', 'Events')
+    recurrence_ids = fields.One2many('calendar.recurrence', 'calendar_id', 'Recurrences')
+
+    calendar_default_privacy = fields.Selection(
+        [('public', 'Public by default'),
+         ('private', 'Private by default'),
+         ('confidential', 'Internal users only'),
+         ('members_only', 'Shared Calendar Members Only')],
+        default=lambda self: self.env['ir.config_parameter'].sudo().get_str('calendar.default_privacy', 'private'),
+        required=True,
+    )
+
+    # All user membership records of this calendar
+    calendar_user_ids = fields.One2many('calendar.user', inverse_name='calendar_id', string='Users', default=_default_calendar_user_ids)
+    # Because we support importing shared calendars through sync, we can have calendars without an owner_id.
+    owner_id = fields.Many2one('res.users', compute='_compute_owner_id')
+    share_user_ids = fields.Many2many('res.users', string='Shared with', compute='_compute_share_user_ids',
+        inverse='_inverse_share_user_ids', domain="[('id', '!=', owner_id), ('share', '=', False)]")
+    # The current user's membership record of this calendar, if any
+    calendar_user_id = fields.Many2one(
+        'calendar.user',
+        string='Current Membership Record',
+        compute='_compute_calendar_user',
+        search='_search_calendar_user_id',
+    )
+    color = fields.Integer(related='calendar_user_id.filter_color', readonly=False)
+    is_primary = fields.Boolean(related='calendar_user_id.is_primary')
+    """
+    Access roles based on Google Calendar. The only roles managed in the base calendar module are:
+        owner - has full access to the calendar, can read/write/delete events and change the calendar settings
+       writer - can read/write/delete non-private events, but cannot edit settings like calendar privacy
+
+    Other roles (reader, freeBusyReader) are managed in the google_calendar module, as they are not relevant
+    to the base calendar module.
+    """
+    user_access_role = fields.Selection(related='calendar_user_id.access_role')
+
+    def write(self, vals):
+        """ Forbid the calendar default privacy update from different users for keeping private events secured. """
+        if ('calendar_default_privacy' in vals and
+                any(self.env.user != calendar.owner_id and not self.env.su for calendar in self)):
+            raise AccessError(_("You are not allowed to change the default privacy of a calendar you do not own."))
+
+        blacklisted_fields = [k for k in vals if k not in self._get_public_writable_fields()]
+        if blacklisted_fields and not self.env.su and any(self.env.user != calendar.owner_id for calendar in self):
+            raise AccessError(_("Only the calendar owner can modify these fields: %(fields)s", fields=", ".join(blacklisted_fields)))
+
+        return super().write(vals)
+
+    @api.depends_context('uid')
+    @api.depends('calendar_user_ids')
+    def _compute_calendar_user(self):
+        """ Gets the calendar user record for the current user, if present."""
+        for calendar in self:
+            calendar.calendar_user_id = calendar.calendar_user_ids.filtered(lambda c: c.user_id == self.env.user)
+
+    @api.model
+    def _search_calendar_user_id(self, operator, value):
+        """ Search is needed to allow using the computed calendar_user_id field for related fields """
+        return [('id', operator, self.env['calendar.user']._search([('calendar_id', 'in', self.ids), ('user_id', '=', self.env.user.id)]))]
+
+    @api.depends('calendar_user_ids.user_id', 'calendar_user_ids.access_role')
+    def _compute_share_user_ids(self):
+        for calendar in self:
+            # Do not include the owner in this list, as we do not want the user to be able to remove themselves.
+            # Doing so could lead to accidental cascade deletion of the calendar
+            calendar.share_user_ids = calendar.calendar_user_ids.filtered(lambda cu: cu.access_role != 'owner').user_id
+
+    def _inverse_share_user_ids(self):
+        for calendar in self:
+            non_owner_calendar_users = calendar.calendar_user_ids.filtered(lambda cu: cu.access_role != 'owner')
+            users_before = non_owner_calendar_users.user_id
+            users_after = calendar.share_user_ids
+            users_to_remove = users_before - users_after
+            users_to_add = users_after - users_before
+
+            if users_to_remove:
+                non_owner_calendar_users.filtered(lambda cu: cu.user_id in users_to_remove).unlink()
+
+            self.env['calendar.user'].create([{
+                'access_role': 'writer',
+                'calendar_id': calendar.id,
+                'name': calendar.name,
+                'user_id': user.id,
+            } for user in users_to_add])
+
+    @api.depends_context('uid')
+    @api.depends('calendar_user_id')
+    def _compute_display_name(self):
+        for calendar in self:
+            if self.env.user == calendar.owner_id:
+                calendar.display_name = calendar.name
+            else:
+                calendar.display_name = calendar.calendar_user_id.name or calendar.name
+
+    @api.depends('calendar_user_ids.access_role', 'calendar_user_ids.user_id')
+    def _compute_owner_id(self):
+        owner_records = self.env['calendar.user'].search([
+            ('access_role', '=', 'owner'),
+            ('calendar_id', 'in', self.ids),
+        ])
+        for calendar in self:
+            calendar.owner_id = owner_records.filtered(lambda l: l.calendar_id.id == calendar.id).user_id.id
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_except_primary(self):
+        if self.calendar_user_ids.filtered('is_primary'):
+            raise UserError(_("A primary calendar cannot be deleted."))
+
+    def add_filter_to_list(self):
+        self.calendar_user_id.is_filter_active = not self.calendar_user_id.is_filter_active
+        self.calendar_user_id.is_filter_checked = True
+
+    def unlink(self):
+        events = self.env['calendar.event'].with_context(active_test=False).search([('calendar_id', 'in', self.ids)])
+        recurrences = self.env['calendar.recurrence'].with_context(active_test=False).search([('calendar_id', 'in', self.ids)])
+        # We do not want to delete the events on Google - disconnect them before deletion
+        events._before_calendar_cascade_unlink()
+        recurrences._before_calendar_cascade_unlink()
+        events.unlink()
+        recurrences.unlink()
+        return super().unlink()
+
+    def _get_public_writable_fields(self):
+        """Sync modules can override this to allow write access to some fields."""
+        return set()

--- a/addons/calendar/models/calendar_event.py
+++ b/addons/calendar/models/calendar_event.py
@@ -30,7 +30,7 @@ from odoo.addons.mail.tools.discuss import Store
 from odoo.tools.intervals import intervals_overlap
 from odoo.tools.translate import _
 from odoo.tools.misc import get_lang, babel_locale_parse
-from odoo.tools import SQL, html2plaintext, html_sanitize, is_html_empty, single_email_re, format_date, format_time
+from odoo.tools import html2plaintext, html_sanitize, is_html_empty, single_email_re, format_date, format_time
 from odoo.exceptions import UserError, ValidationError
 
 _logger = logging.getLogger(__name__)
@@ -166,6 +166,10 @@ class CalendarEvent(models.Model):
 
     # description
     name = fields.Char('Meeting Subject', required=True, copy=True)
+    calendar_id = fields.Many2one('calendar.calendar', string='Calendar', index='btree',
+        compute='_compute_calendar_id', store=True, readonly=False, ondelete='cascade',
+        domain=lambda self: self._get_calendar_id_domain())
+    calendar_color = fields.Integer(related='calendar_id.color')
     description = fields.Html('Description',
         help="""When synchronization with an external calendar is active, this description is synchronized \
         with the one of the associated meeting in that external calendar. Any update will be propagated there \
@@ -190,13 +194,15 @@ class CalendarEvent(models.Model):
     privacy = fields.Selection(
         [('public', 'Public'),
          ('private', 'Private'),
-         ('confidential', 'Only internal users')], 'Privacy',
+         ('confidential', 'Only internal users'),
+         ('members_only', 'Shared Calendar Members Only')], 'Privacy',
         help="People to whom this event will be visible.")
     privacy_placeholder = fields.Char(compute='_compute_privacy_placeholder')
     effective_privacy = fields.Selection(
-        [('public', 'Public'), ('private', 'Private'), ('confidential', 'Only internal users')],
+        [('public', 'Public'), ('private', 'Private'),
+         ('confidential', 'Only internal users'), ('members_only', 'Shared Calendar Members Only')],
         'Effective Privacy', help="Whether the event is private, considering the user privacy",
-        compute="_compute_effective_privacy"
+        compute="_compute_effective_privacy", search="_search_effective_privacy"
     )
     show_as = fields.Selection(
         [('free', 'Available'),
@@ -332,6 +338,18 @@ class CalendarEvent(models.Model):
         for event in self.filtered(lambda event: event.allday):
             event.show_as = 'free'
 
+    @api.depends('user_id')
+    def _compute_calendar_id(self):
+        for event in self:
+            if event.user_id != event.calendar_id.owner_id:
+                event.calendar_id = event.user_id._find_or_create_primary_calendar() if event.user_id else False
+
+    def _get_calendar_id_domain(self):
+        return Domain("calendar_user_ids", "any", [
+            ("user_id", "=", self.env.uid),
+            ("access_role", "in", ["writer", "owner"]),
+        ])
+
     @api.depends("attendee_ids")
     def _compute_should_show_status(self):
         for event in self:
@@ -379,10 +397,13 @@ class CalendarEvent(models.Model):
                 'awaiting_count': attendees_count - accepted_count - declined_count - tentative_count
             })
 
-    @api.depends('partner_ids')
+    @api.depends('partner_ids', 'calendar_id')
     @api.depends_context('uid')
     def _compute_user_can_edit(self):
         for event in self:
+            if event.calendar_id and event.calendar_id.id in self.env.user.writable_calendar_ids.ids:
+                event.user_can_edit = True
+                continue
             # By default, only current attendees, the organizer and the creator can edit the event.
             editor_candidates = set(event.partner_ids.user_ids + event.user_id + (event.create_uid or self.env.user))
             # Right before saving the event, old partners must be able to save changes.
@@ -400,10 +421,41 @@ class CalendarEvent(models.Model):
                 lambda a: not (a.email and single_email_re.match(a.email))
             )
 
-    @api.depends('privacy', 'user_id')
+    @api.depends('privacy', 'calendar_id.calendar_default_privacy', 'user_id.primary_calendar_id.calendar_default_privacy')
     def _compute_effective_privacy(self):
         for event in self:
-            event.effective_privacy = event.privacy or event.sudo().user_id.calendar_default_privacy
+            event.effective_privacy = (event.privacy
+                                       or event.sudo().calendar_id.calendar_default_privacy
+                                       or event.user_id._find_or_create_primary_calendar().calendar_default_privacy if event.user_id else 'public')
+
+    def _search_effective_privacy(self, operator, value):
+        if operator not in ('in', 'not in'):
+            return NotImplemented
+
+        privacy_field_domain = Domain('privacy', operator, value)
+        # If no privacy is set on the event, fallback to the calendar default privacy.
+        calendar_default_privacy_domain = Domain.AND([
+            Domain('privacy', '=', False), Domain('calendar_id.calendar_default_privacy', operator, value),
+        ])
+        # If no privacy is set on the event nor on the calendar, fallback to the user default privacy.
+        user_primary_calendar_default_privacy_domain = Domain.AND([
+            Domain('privacy', '=', False),
+            Domain('calendar_id.calendar_default_privacy', '=', False),
+            Domain('user_id.primary_calendar_id.calendar_default_privacy', operator, value),
+        ])
+        # If no privacy is set on the event nor on the calendar or the user, fallback to the public privacy.
+        public_domain_fallback = Domain.AND([
+            Domain('user_id', '=', False),
+            Domain('privacy', '=', False),
+            Domain('calendar_id.calendar_default_privacy', '=', False),
+        ])
+
+        return Domain.OR([
+            privacy_field_domain,
+            calendar_default_privacy_domain,
+            user_primary_calendar_default_privacy_domain,
+            public_domain_fallback,
+        ])
 
     @api.depends('effective_privacy')
     def _compute_privacy_placeholder(self):
@@ -953,9 +1005,6 @@ class CalendarEvent(models.Model):
         if any(vals in self._get_recurrent_fields() for vals in values) and not (update_recurrence or values.get('recurrency')):
             raise UserError(_('Unable to save the recurrence with "This Event"'))
 
-        # Check the privacy permissions of the events whose organizer is different from the current user.
-        self.filtered(lambda ev: ev.user_id and self.env.user != ev.user_id)._check_calendar_privacy_write_permissions()
-
         if set(values) == {'videocall_channel_id'}:
             # Attaching the Discuss channel of a meeting is no business change: it must not go
             # through the recurrence machinery below, which would recreate the occurrences from
@@ -965,7 +1014,7 @@ class CalendarEvent(models.Model):
         update_alarms = False
         update_time = False
         self._set_videocall_location([values])
-        if 'partner_ids' in values:
+        if values.get('partner_ids'):
             if values['partner_ids'] and isinstance(values['partner_ids'][0], int):
                 # a plain list of ids stands for a `Command.set`, as `_attendees_values` assumes too
                 values['partner_ids'] = [Command.set(values['partner_ids'])]
@@ -1067,26 +1116,36 @@ class CalendarEvent(models.Model):
 
         return True
 
-    def _check_calendar_privacy_write_permissions(self):
-        """
-        Checks if current user can write on the events, raising UserError when the event is private.
-        We need to manually call the default Access Error because we can't add an access rule for checking
-        the calendar defaut privacy of an user from a 'calendar.event' record, since it is a res.users field.
-        Otherwise we would have to create a new computed field on that model, which we don't want.
-        """
-        if not self.env.su:
-            for event in self:
-                if event._check_private_event_conditions():
-                    raise event._make_access_error_message('write', Domain.TRUE)  # noqa: EM101
-
     def _check_private_event_conditions(self):
-        """ Checks if the event is private, returning True if the conditions match and False otherwise. """
+        """ Returning True if the event should be considered private and False otherwise.
+        This check is used so that we can fetch private events and display 'Busy' slots in the calendar,
+        without compromising any of the private fields (see usage in _fetch_query). These conditions
+        should be kept in sync with the conditions defined by the calendar_event_rule_private ACL rule.
+        """
         self.ensure_one()
-        event_is_private = self.privacy == 'private'
-        calendar_is_private = not self.privacy and self.sudo().user_id.calendar_default_privacy == 'private'
-        user_is_not_partner = self.user_id.id != self.env.uid and self.env.user.partner_id not in self.partner_ids
-        user_is_not_creator = self.env.user != (self.create_uid or self.env.user)  # check if user has created the event
-        return (event_is_private or calendar_is_private) and user_is_not_partner and user_is_not_creator
+        if self.effective_privacy == 'public':
+            return False
+
+        # Confidential events are only visible to internal users
+        if self.effective_privacy == 'confidential' and self.env.user._is_internal():
+            return False
+
+        # Organizer, attendees, and event creator should be able to access the event
+        if (self.user_id.id == self.env.uid or
+                self.env.user == self.create_uid or
+                self.env.user.partner_id in self.partner_ids):
+            return False
+
+        # The calendar owner has access to all events in their calednar
+        if self.calendar_id and self.calendar_id.user_access_role == 'owner':
+            return False
+
+        # Events marked as members_only should be shown to all members of the calendar
+        if self.effective_privacy == 'members_only':
+            user_has_calendar_access = self.calendar_id and self.calendar_id.user_access_role in ['owner', 'writer']
+            return not user_has_calendar_access
+
+        return True
 
     @api.depends('privacy', 'user_id')
     def _compute_display_name(self):
@@ -1370,17 +1429,35 @@ class CalendarEvent(models.Model):
         return Domain("start", "<=", end_of_day) & Domain("stop", ">=", start_of_day)
 
     def _get_default_privacy_domain(self):
-        # Sub query user settings from calendars that are not private ('public' and 'confidential').
-        public_calendars_settings = self.env['res.users.settings'].sudo()._search([('calendar_default_privacy', '!=', 'private')]).select(SQL('user_id'))
-        # display public, confidential events and events with default privacy when owner's default privacy is not private
-        return ['|', '|',
-            ('privacy', 'in', ['public', 'confidential']),
+        # Sub query user settings from calendars that are not private
+        calendar_domain = Domain(['|',
+            ('calendar_default_privacy', '=', 'public'),
+            '&',
+                ('calendar_default_privacy', '=', 'members_only'),
+                ('calendar_user_id', '!=', False)])
+
+        if self.env.user._is_internal():
+            calendar_domain = Domain.OR([calendar_domain, Domain('calendar_default_privacy', '=', 'confidential')])
+
+        accessible_calendars = self.env['calendar.calendar'].sudo()._search(calendar_domain)
+
+        # Get public events, events the user is attending or organizing, or events the user has calendar access to
+        event_domain = Domain(['|', '|', '|',
+            ('privacy', '=', 'public'),
             ('user_id', '=', self.env.user.id),
+            ('partner_ids', 'in', self.env.user.partner_id.id),
             '&',
-                ('privacy', '=', False),
+                ('privacy', '=', False),  # fallback to calendar privacy
                 '|',
-                    ('user_id', '=', False),
-                    ('user_id', 'in', public_calendars_settings)]
+                    '&',
+                        ('user_id', '=', False),
+                        ('calendar_id', '=', False),
+                    ('calendar_id', 'in', accessible_calendars)])
+
+        if self.env.user._is_internal():
+            event_domain = Domain.OR([event_domain, Domain('privacy', '=', 'confidential')])
+
+        return event_domain
 
     def _is_event_over(self):
         """Check if the event is over. This method is used to check if the event
@@ -1398,6 +1475,10 @@ class CalendarEvent(models.Model):
         # For timed events
         return self.stop and self.stop < now
 
+    def _before_calendar_cascade_unlink(self):
+        """Override this method to perform actions before the cascade unlink of the event when deleting a calendar."""
+        return
+
     # ------------------------------------------------------------
     # ACTIONS
     # ------------------------------------------------------------
@@ -2061,7 +2142,7 @@ class CalendarEvent(models.Model):
     def _get_public_fields(self):
         return self._get_recurrent_fields() | self._get_time_fields() | self._get_custom_fields() | {
             'id', 'active', 'allday',
-            'duration', 'user_id', 'interval', 'partner_id',
+            'duration', 'user_id', 'interval', 'partner_id', 'calendar_id',
             'count', 'rrule', 'recurrence_id', 'show_as', 'privacy', 'create_uid'}
 
     @api.model

--- a/addons/calendar/models/calendar_recurrence.py
+++ b/addons/calendar/models/calendar_recurrence.py
@@ -96,8 +96,9 @@ class CalendarRecurrence(models.Model):
 
     name = fields.Char(compute='_compute_name', store=True)
     base_event_id = fields.Many2one(
-        'calendar.event', ondelete='set null', copy=False)  # store=False ?
+        'calendar.event', ondelete='set null', index='btree', copy=False)  # store=False ?
     calendar_event_ids = fields.One2many('calendar.event', 'recurrence_id')
+    calendar_id = fields.Many2one(related='base_event_id.calendar_id', string='Calendar')
     event_tz = fields.Selection(
         _tz_get, string='Timezone',
         default=lambda self: self.env.context.get('tz') or self.env.user.tz)
@@ -643,3 +644,7 @@ class CalendarRecurrence(models.Model):
             (event.stop_date < today if event.allday else event.stop < now)
             for event in self.calendar_event_ids
         )
+
+    def _before_calendar_cascade_unlink(self):
+        """Override this method to perform actions before the cascade unlink of the recurrence when deleting a calendar."""
+        return

--- a/None
+++ b/addons/calendar/models/calendar_user.py
@@ -0,0 +1,48 @@
+from random import randint
+
+from odoo import api, fields, models, _
+from odoo.exceptions import AccessError
+
+
+class CalendarUser(models.Model):
+    _name = 'calendar.user'
+    _description = 'Calendar User Manager'
+
+    def _get_default_color(self):
+        return randint(1, 11)
+
+    _unique_user_per_calendar = models.UniqueIndex('(user_id, calendar_id)')
+    _single_primary_calendar_per_user = models.UniqueIndex('(user_id) WHERE is_primary')
+
+    calendar_id = fields.Many2one('calendar.calendar', string='Calendar', ondelete='cascade', index='btree', required=True, readonly=True)
+    user_id = fields.Many2one('res.users', string='User', required=True, readonly=True)
+    is_primary = fields.Boolean('Primary', readonly=True)
+    name = fields.Char('Label')
+
+    # Access roles matching those of Google Calendar
+    access_role = fields.Selection([
+        ('owner', 'Owner'),
+        ('writer', 'Write'),
+    ], required=True, readonly=True)
+
+    # Filter values
+    filter_color = fields.Integer(string='Color', default=_get_default_color)
+    is_filter_active = fields.Boolean('Shown in Filters List', default=True)
+    is_filter_checked = fields.Boolean('Filter Enabled', default=True)
+
+    def write(self, vals):
+        blacklisted_fields = [k for k in vals if k not in self._get_writeable_fields()]
+        if blacklisted_fields and not self.env.su:
+            raise AccessError(_("These fields cannot be modified: %(fields)s", fields=", ".join(blacklisted_fields)))
+
+        return super().write(vals)
+
+    def unlink(self):
+        calendars_to_remove = self.calendar_id.filtered(lambda cal: cal.calendar_user_ids <= self)
+        result = super().unlink()
+        calendars_to_remove.sudo().unlink()
+        return result
+
+    @api.model
+    def _get_writeable_fields(self):
+        return {'is_filter_active', 'is_filter_checked', 'filter_color', 'name'}

--- a/addons/calendar/models/res_users.py
+++ b/addons/calendar/models/res_users.py
@@ -3,22 +3,61 @@
 import datetime
 from zoneinfo import ZoneInfo
 
-from odoo import api, fields, models, modules, _
-from odoo.exceptions import AccessError
+from odoo import api, Command, fields, models, modules, _
 from odoo.fields import Domain
 
 
 class ResUsers(models.Model):
     _inherit = 'res.users'
 
-    calendar_default_privacy = fields.Selection(
-        [('public', 'Public by default'),
-         ('private', 'Private by default'),
-         ('confidential', 'Internal users only')],
-        compute="_compute_calendar_default_privacy",
-        inverse="_inverse_calendar_res_users_settings",
-        user_writeable=True,
-    )
+    calendar_user_ids = fields.One2many('calendar.user', 'user_id')
+    calendar_ids = fields.One2many('calendar.calendar', compute='_compute_calendar_ids', readonly=False, user_writeable=True)
+    owned_calendar_ids = fields.One2many('calendar.calendar', compute='_compute_calendar_ids')
+    writable_calendar_ids = fields.One2many('calendar.calendar', compute='_compute_calendar_ids')
+    primary_calendar_id = fields.Many2one('calendar.calendar', compute='_compute_primary_calendar', store=True)
+
+    @api.depends('calendar_user_ids.is_primary', 'calendar_user_ids.access_role', 'calendar_user_ids.calendar_id')
+    def _compute_primary_calendar(self):
+        for user in self:
+            user.primary_calendar_id = user.calendar_user_ids.filtered(
+                lambda l: l.is_primary and l.access_role == 'owner').calendar_id
+
+    @api.depends('calendar_user_ids')
+    def _compute_calendar_ids(self):
+        calendar_user_records = self.env['calendar.user'].sudo().search([('user_id', 'in', self.ids)])
+        for user in self:
+            user.calendar_ids = calendar_user_records.filtered(lambda l: l.user_id.id == user.id).calendar_id
+            user.owned_calendar_ids = calendar_user_records.filtered(
+                lambda l: l.user_id.id == user.id and l.access_role == 'owner').calendar_id
+            user.writable_calendar_ids = calendar_user_records.filtered(
+                lambda l: l.user_id.id == user.id and l.access_role in ('owner', 'writer')).calendar_id
+
+    @api.model
+    def _find_or_create_primary_calendar(self):
+        self.ensure_one()
+        if self.primary_calendar_id:
+            return self.primary_calendar_id
+
+        self._generate_primary_calendar()
+        # Field should be recomputed after the creation of the primary calendar
+        return self.primary_calendar_id
+
+    def _generate_primary_calendar(self):
+        self.env['calendar.calendar'].sudo().create([
+            {
+                'name': user.display_name,
+                'calendar_default_privacy': self.env['ir.config_parameter'].sudo().get_str('calendar.default_privacy', 'public'),
+                'calendar_user_ids': [Command.create({
+                    'user_id': user.id,
+                    'is_primary': True,
+                    'access_role': 'owner',
+                    'is_filter_active': True,
+                    'is_filter_checked': True,
+                    'name': user.display_name,
+                })],
+            }
+            for user in self
+        ])
 
     def get_selected_calendars_partner_ids(self, include_user=True):
         """
@@ -38,58 +77,17 @@ class ResUsers(models.Model):
             partner_ids += [self.env.user.partner_id.id]
         return partner_ids
 
-    @api.model
-    def _default_user_calendar_default_privacy(self):
-        """ Get the calendar default privacy from the Default User Template, set public as default. """
-        return self.env['ir.config_parameter'].sudo().get_str('calendar.default_privacy', 'public')
-
     @api.model_create_multi
     def create(self, vals_list):
-        """ Set the calendar default privacy as the same as Default User Template when defined. """
-        default_privacy = self._default_user_calendar_default_privacy()
-        # Update the dictionaries in vals_list with the calendar default privacy.
-        for vals_dict in vals_list:
-            if not vals_dict.get('calendar_default_privacy'):
-                vals_dict.update(calendar_default_privacy=default_privacy)
-
-        res = super().create(vals_list)
-        return res
-
-    def write(self, vals):
-        """ Forbid the calendar default privacy update from different users for keeping private events secured. """
-        privacy_update = 'calendar_default_privacy' in vals
-        if privacy_update and self != self.env.user:
-            raise AccessError(_("You are not allowed to change the calendar default privacy of another user due to privacy constraints."))
-        return super().write(vals)
-
-    @api.depends("res_users_settings_id.calendar_default_privacy")
-    def _compute_calendar_default_privacy(self):
-        """
-        Compute the calendar default privacy of the users, pointing to its ResUsersSettings.
-        When any user doesn't have its setting from ResUsersSettings defined, fallback to Default User Template's.
-        """
-        fallback_default_privacy = 'public'
-        # sudo: any user has access to other users calendar_default_privacy setting
-        if any(not user.sudo().res_users_settings_id.calendar_default_privacy for user in self):
-            fallback_default_privacy = self._default_user_calendar_default_privacy()
-
-        for user in self:
-            user.calendar_default_privacy = user.sudo().res_users_settings_id.calendar_default_privacy or fallback_default_privacy
-
-    def _inverse_calendar_res_users_settings(self):
-        """
-        Updates the values of the calendar fields in 'res_users_settings_ids' to have the same values as their related
-        fields in 'res.users'. If there is no 'res.users.settings' record for the user, then the record is created.
-        """
-        for user in self.filtered(lambda user: user._is_internal()):
-            settings = self.env["res.users.settings"].sudo()._find_or_create_for_user(user)
-            configuration = {field: user[field] for field in self._get_user_calendar_configuration_fields()}
-            settings.sudo().update(configuration)
-
-    @api.model
-    def _get_user_calendar_configuration_fields(self) -> list[str]:
-        """ Return the list of configurable fields for the user related to the res.users.settings table. """
-        return ['calendar_default_privacy']
+        users = super().create(vals_list)
+        users.filtered(lambda u: not u.share)._generate_primary_calendar()
+        return users
+
+    def unlink(self):
+        # Normal ondelete=cascade does not go through unlink() overrides, which we need to remove
+        # the calendar if this was its last user. Otherwise, we would end up with orphaned calendars.
+        self.calendar_user_ids.unlink()
+        return super().unlink()
 
     def _systray_get_calendar_event_domain(self):
         # Determine the domain for which the users should be notified. This method sends notification to

--- a/addons/calendar/models/res_users_settings.py
+++ b/addons/calendar/models/res_users_settings.py
@@ -1,24 +1,9 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models
+from odoo import fields, models
 
 
 class ResUsersSettings(models.Model):
     _inherit = "res.users.settings"
 
-    # Calendar module settings.
-    calendar_default_privacy = fields.Selection(
-        [('public', 'Public'),
-         ('private', 'Private'),
-         ('confidential', 'Only internal users')],
-        'Calendar Default Privacy', default='public', required=True,
-        store=True, readonly=False, help="Default privacy setting for whom the calendar events will be visible."
-    )
     calendar_show_activities = fields.Boolean(string='Show Activities in Calendar', default=True)
-    calendar_hide_my = fields.Boolean(string='Hide My Calendar from Calendar')
-
-    @api.model
-    def _get_fields_blacklist(self):
-        """ Get list of calendar fields that won't be formatted in session_info. """
-        calendar_fields_blacklist = ['calendar_default_privacy']
-        return super()._get_fields_blacklist() + calendar_fields_blacklist
